### DISCUSSION TITLE
May I know how to use this library to convert to TIFF to android bitmap?

### DISCUSSION THREAD

**Discussion by Jimex** [ngageoint/tiff-java#43] [Q&A] [answered]
May I know how to use this library to convert to TIFF to android bitmap?

Thank you.

**Comment by bosborn** [comment]
Here is an example of converting a RBG tiff ([example](https://github.com/ngageoint/tiff-java/raw/master/src/test/resources/rgb.tiff)) to a Bitmap

```
TIFFImage tiffImage = TiffReader.readTiff(/* file, bytes, reader, or stream */);
FileDirectory fileDirectory = tiffImage.getFileDirectory();

Rasters rasters = fileDirectory.readRasters();
int width = rasters.getWidth();
int height = rasters.getHeight();

Bitmap bitmap = null;

switch(fileDirectory.getPhotometricInterpretation()){

    case TiffConstants.PHOTOMETRIC_INTERPRETATION_RGB:

        bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);

        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                Number[] pixel = rasters.getPixel(x, y);
                int color = Color.rgb(pixel[0].intValue(), pixel[1].intValue(), pixel[2].intValue());
                bitmap.setPixel(x, y, color);
            }
        }

        break;

    default:

}
```

**Reply by Jimex** [reply]
Thank you for the reply!
I have tried it and it works, however, the image quality becomes bad after conversion.

**Before**
![image](https://user-images.githubusercontent.com/3145858/147997026-5a177a67-5761-47a7-8a67-89bf0e027037.png)


**After**
![image](https://user-images.githubusercontent.com/3145858/147996948-91fc7cec-99c4-4835-9e24-f2094d24334e.png)

**Reply by bosborn** [reply]
It probably just depends on how the TIFF is encoded, can you post the actual TIFF file?

**Reply by Jimex** [reply]
Yes, I think it might be related to the foramt of the TIFF too, I  tried decode the tiff file by using UTIF.js and it could be display well.

Here is the actual TIFF file, thank you.

[test-tiff.zip](https://github.com/ngageoint/tiff-java/files/7807399/test-tiff.zip)

**Comment by bosborn** [comment, marked as answer]
That tiff file has predictor compression (in addition to LZW).  Predictor decoding is currently not implemented.

Until implemented, here is a workaround that worked for your image:

```
TIFFImage tiffImage = TiffReader.readTiff(/* file, bytes, reader, or stream */);
FileDirectory fileDirectory = tiffImage.getFileDirectory();

Rasters rasters = fileDirectory.readRasters();
int width = rasters.getWidth();
int height = rasters.getHeight();

Bitmap bitmap = null;

switch(fileDirectory.getPhotometricInterpretation()){

    case TiffConstants.PHOTOMETRIC_INTERPRETATION_RGB:

        Integer predictor = fileDirectory.getIntegerEntryValue(FieldTagType.Predictor);
        boolean horizontalPredictor = predictor != null && predictor == 2;

        bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);

        for (int y = 0; y < height; y++) {
            Number[] previousPixel = null;
            for (int x = 0; x < width; x++) {
                Number[] pixel = rasters.getPixel(x, y);
                if(previousPixel != null){
                    for(int sample = 0; sample < pixel.length; sample++){
                        pixel[sample] = (pixel[sample].intValue() + previousPixel[sample].intValue()) % 256;
                    }
                }
                int color = Color.rgb(pixel[0].intValue(), pixel[1].intValue(), pixel[2].intValue());
                bitmap.setPixel(x, y, color);
                if(horizontalPredictor) {
                    previousPixel = pixel;
                }
            }
        }

        break;

    default:

}
```
![Screen Shot 2022-01-05 at 6 34 20 AM](https://user-images.githubusercontent.com/11667761/148226233-b730ae41-bec8-40f6-a452-79d13cc2d09f.png)

**Reply by Jimex** [reply]
Amazing, it works like a charm now, and the speed is 5x faster than the previous one.

Thank you very much!

**Reply by bosborn** [reply]
After updating to [version 3.0.0](https://github.com/ngageoint/tiff-java/releases/tag/3.0.0), you should be able to revert back to [the previous answer](https://github.com/ngageoint/tiff-java/discussions/43#discussioncomment-1902423).

**Reply by Jimex** [reply]
Thank you for the update, I have updated my code accordingly.

By the way, does this library still cannot support the grayscale TIFF? I got the following error when I try to convert some black-and-white tiff to bitmap.

_W/System.err: mil.nga.tiff.util.TiffException: Unsupported field type for sample format: 1, bits per sample: 4
W/System.err:     at mil.nga.tiff.FieldType.getFieldType(FieldType.java:185)
W/System.err:     at mil.nga.tiff.FileDirectory.getFieldTypeForSample(FileDirectory.java:1383)
W/System.err:     at mil.nga.tiff.FileDirectory.readRasters(FileDirectory.java:1212)
W/System.err:     at mil.nga.tiff.FileDirectory.readRasters(FileDirectory.java:1067)
W/System.err:     at mil.nga.tiff.FileDirectory.readRasters(FileDirectory.java:1019)
W/System.err:     at mil.nga.tiff.FileDirectory.readRasters(FileDirectory.java:998)_

The file format as below:
Image:
  Format: TIFF (Tagged Image File Format)
  Mime type: image/tiff
  Class: PseudoClass
  Geometry: 2115x3123+0+0
  Resolution: 300x300
  Print size: 7.05x10.41
  Units: PixelsPerInch
  Colorspace: sRGB
  Type: Grayscale
  Base type: Palette
  Endianness: LSB
  Depth: 4/8-bit
  Channel depth:
    Red: 8-bit
    Green: 8-bit
    Blue: 8-bit
  Channel statistics:
    Pixels: 6605145
    Red:
      min: 0.0588235  (0.00392157)
      max: 15 (1)
      mean: 13.1515 (0.876767)
      median: 15 (1)
      standard deviation: 4.68201 (0.312134)
      kurtosis: 3.28602
      skewness: -2.26044
      entropy: 0.242478
    Green:
      min: 0.0588235  (0.00392157)
      max: 15 (1)
      mean: 13.1515 (0.876767)
      median: 15 (1)
      standard deviation: 4.68201 (0.312134)
      kurtosis: 3.28602
      skewness: -2.26044
      entropy: 0.242478
    Blue:
      min: 0.0588235  (0.00392157)
      max: 15 (1)
      mean: 13.1515 (0.876767)
      median: 15 (1)
      standard deviation: 4.68201 (0.312134)
      kurtosis: 3.28602
      skewness: -2.26044
      entropy: 0.242478
  Image statistics:
    Overall:
      min: 0.0588235  (0.00392157)
      max: 15 (1)
      mean: 13.1515 (0.876767)
      median: 15 (1)
      standard deviation: 4.68201 (0.312134)
      kurtosis: 3.28602
      skewness: -2.26044
      entropy: 0.242478
  Colors: 16
  Histogram:
    594702: (1,1,1) #010101 srgb(1,1,1)
    57392: (23,23,23) #171717 grey9
    38617: (39,39,39) #272727 srgb(39,39,39)
    34549: (55,55,55) #373737 srgb(55,55,55)
    29419: (71,71,71) #474747 grey28
    27644: (87,87,87) #575757 grey34
    26836: (104,104,104) #686868 srgb(104,104,104)
    22060: (120,120,120) #787878 grey47
    21871: (135,135,135) #878787 grey53
    23120: (152,152,152) #989898 srgb(152,152,152)
    21422: (168,168,168) #A8A8A8 grey66
    20607: (183,183,183) #B7B7B7 srgb(183,183,183)
    20740: (200,200,200) #C8C8C8 srgb(200,200,200)
    19368: (215,215,215) #D7D7D7 srgb(215,215,215)
    24403: (231,231,231) #E7E7E7 srgb(231,231,231)
    5622395: (255,255,255) #FFFFFF white
  Colormap entries: 16
  Colormap:
    0: (1,1,1,1) #010101FF srgba(1,1,1,1)
    1: (23,23,23,1) #171717FF grey9
    2: (39,39,39,1) #272727FF srgba(39,39,39,1)
    3: (55,55,55,1) #373737FF srgba(55,55,55,1)
    4: (71,71,71,1) #474747FF grey28
    5: (87,87,87,1) #575757FF grey34
    6: (104,104,104,1) #686868FF srgba(104,104,104,1)
    7: (120,120,120,1) #787878FF grey47
    8: (135,135,135,1) #878787FF grey53
    9: (152,152,152,1) #989898FF srgba(152,152,152,1)
    10: (168,168,168,1) #A8A8A8FF grey66
    11: (183,183,183,1) #B7B7B7FF srgba(183,183,183,1)
    12: (200,200,200,1) #C8C8C8FF srgba(200,200,200,1)
    13: (215,215,215,1) #D7D7D7FF srgba(215,215,215,1)
    14: (231,231,231,1) #E7E7E7FF srgba(231,231,231,1)
    15: (255,255,255,1) #FFFFFFFF white
  Rendering intent: Perceptual
  Gamma: 0.454545
  Chromaticity:
    red primary: (0.64,0.33)
    green primary: (0.3,0.6)
    blue primary: (0.15,0.06)
    white point: (0.3127,0.329)
  Matte color: grey74
  Background color: white
  Border color: srgb(223,223,223)
  Transparent color: none
  Interlace: None
  Intensity: Undefined
  Compose: Over
  Page geometry: 2115x3123+0+0
  Dispose: Undefined
  Iterations: 0
  Compression: LZW
  Orientation: TopLeft
  Properties:
    date:create: 2022-01-15T10:37:22+00:00
    date:modify: 2022-01-15T10:37:28+00:00
    signature: 0e384137334be05c21be0778319e19422570a68143ed679f10a113d434e0b125
    tiff:alpha: unspecified
    tiff:endian: lsb
    tiff:photometric: palette
    tiff:rows-per-strip: 3128
  Artifacts:
    verbose: true
  Tainted: False
  Filesize: 525434B
  Number pixels: 6.60515M
  Pixels per second: 72.118MP
  User time: 0.094u
  Elapsed time: 0:01.091

**Reply by bosborn** [reply]
It does support grayscale images, but not 4 bits (half a byte) per sample.
Supports 8, 16, or 32 bits (1, 2, or 4 bytes) per sample for unsigned integers.

**Reply by Jimex** [reply]
Understood, thank you for the clarification.
